plugins/connection/nsenter.py

```diff
@@ -0,0 +1,245 @@
+                  the target hosts, which is why this feature is disabled by default.
+            env:
+                - name: ANSIBLE_PIPELINING
+            ini:
+                - section: defaults
+                  key: pipelining
+            type: boolean
+            vars:
+                - name: ansible_pipelining
+    notes:
+        - The remote user is ignored; this plugin always runs as root.
+        - "This plugin requires the Ansible controller container to be launched in the following way:"
+        - (1) The container image contains the nsenter program;
+        - (2) The container is launched in privileged mode;
+        - (3) The container is launched in the host's PID namespace (i.e. C(--pid host)).
+'''
+
+import os
+import pty
+import shutil
+import subprocess
+import fcntl
+
+import ansible.constants as C
+from ansible.errors import AnsibleError, AnsibleFileNotFound
+from ansible.module_utils.compat import selectors
+from ansible.module_utils.six import text_type, binary_type
+from ansible.module_utils._text import to_bytes, to_native, to_text
+from ansible.plugins.connection import ConnectionBase
+from ansible.utils.display import Display
+from ansible.utils.path import unfrackpath
+
+display = Display()
+
+
+class Connection(ConnectionBase):
+    '''Connections to a container host using nsenter
+    '''
+
+    transport = 'nsenter'
+    has_pipelining = True
+
+    def __init__(self, *args, **kwargs):
+        super(Connection, self).__init__(*args, **kwargs)
+        self.cwd = None
+        self._nsenter_pid = self.get_option("nsenter_pid")
+
+    def _connect(self):
+        # Because nsenter requires very high privileges, our remote user
+        # is always assumed to be root.
+        self._play_context.remote_user = "root"
+
+        if not self._connected:
+            display.vvv(
+                u"ESTABLISH NSENTER CONNECTION FOR USER: {0}".format(
+                    self._play_context.remote_user
+                ),
+                host=self._play_context.remote_addr,
+            )
+            self._connected = True
+        return self
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.debug("in nsenter.exec_command()")
+
+        executable = C.DEFAULT_EXECUTABLE.split()[0] if C.DEFAULT_EXECUTABLE else None
+
+        if not os.path.exists(to_bytes(executable, errors='surrogate_or_strict')):
+            raise AnsibleError("failed to find the executable specified %s."
+                               " Please verify if the executable exists and re-try." % executable)
+
+        # Rewrite the provided command to prefix it with nsenter
+        if isinstance(cmd, (text_type, binary_type)):
+            nsenter_cmd = "nsenter --all --preserve-credentials --target={0} -- ".format(self._nsenter_pid)
+            cmd = to_bytes(nsenter_cmd + cmd)
```

Comment:
```suggestion
            cmd = to_bytes(nsenter_cmd) + to_bytes(cmd)
```
Otherwise you can end up with funny string contents when `nsenter_cmd` and `cmd` are not of the same type.